### ISSUE TITLE
Binding patterns in parameter lists of function types

### ISSUE THREAD

**Issue by nickie** [Microsoft/TypeScript#6840]
I believe that the following two variable declarations are consistent with the specifications ([3.9.2.2](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3922-parameter-list)).  They define two functions of a single parameter, which is a list of three things.  Notice that a binding pattern is used for destructuring this list.

```TypeScript
var y: ([a, b, c]) => any;
var z: ([a, b, c] : number[]) => any;
```

Both declarations are disallowed by the implementation.  The error messages suggest that `[a, b, c]` is immediately assumed to be a (parenthesized) tuple type, as would be the case in:

```TypeScript
interface a {};
interface b {};
interface c {};

var x: ([a, b, c]);
```

**Comment by sandersn** [comment]
Looks to me like it's failing at the parser level.
I get very wrong error messages when I add a type too.

```ts
let x: ([a,b,c]: [any,any,any]) => any;
```

**Comment by ahejlsberg** [comment]
Yup, that doesn't look right. I will get it fixed.

**Comment by ahejlsberg** [comment]
@nickie Thanks for reporting. Fix now in master.
